### PR TITLE
Circuit breakers

### PR DESCRIPTION
**Pull request by lmineiro**
And some metrics for load tests

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,29 @@
+SHELL := /bin/bash
+
+GOOS ?= linux
+GOARCH ?= amd64
+
+SOURCES = $(shell find $(ROOT_DIR) -name "*.go")
+TARGET_NAME = z-skipper
```
Comment (aryszka) [inline Makefile]: cross branding? :P

Comment (lmineiro) [inline Makefile]: Oops ;-)

File: breaker/breaker.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+)
+
+func Do(name string, url string) (resp *http.Response, err error) {
+	return DoWithFallback(name, url, nil)
+}
+
+func DoWithFallback(name string, url string, f func(error) error) (resp *http.Response, err error) {
+	err = hystrix.Do(name, func() error {
+		start := time.Now()
+		if resp, err = Get(url); err == nil {
```
Comment (aryszka) [inline breaker/breaker.go]: if the client in the same package allows non-default client instances, then i think the breaker should allow it, too. this Get calls always the default client. it's cool for the app i guess, but the package would be nicer if the breaker could use a non-default client.

Comment (lmineiro) [inline breaker/breaker.go]: Right now it's not needed. I guess they should be in separate packages. I'll consider that. Thanks

File: breaker/breaker.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+			measureRequest(start, fmt.Sprintf("planb.breaker.%s", name))
+		}
+		return err
+	}, f)
+	return
+}
+
+func measureRequest(start time.Time, key string) {
+	t := metrics.DefaultRegistry.GetOrRegister(key, metrics.NewTimer).(metrics.Timer)
```
Comment (aryszka) [inline breaker/breaker.go]: my previous comment probably applies to the defaultregistry, too. Maybe, there could be a breaker object, that encapsulates the client, and the metrics, and if those are not set, only then uses the default.

Comment (lmineiro) [inline breaker/breaker.go]: Sounds good

File: handlers/tokeninfo/jwt/handler.go [hunk 4]
```diff
@@ -45,19 +52,30 @@ func (h *jwtHandler) Match(r *http.Request) bool {
...
 func (h *jwtHandler) validateToken(req *http.Request) (*TokenInfo, error) {
+	start := time.Now()
 	token, err := jwt.ParseFromRequest(req, jwtValidator(h.keyLoader))
+	if err == nil {
+		measureRequest(start, fmt.Sprintf("planb.tokeninfo.jwt.validation.%s", token.Method.Alg()))
+	}
 	if err == nil && token.Valid {
 		return newTokenInfo(token, time.Now())
 	} else {
-		log.Infof("Failed to validate token: %s", err)
```
Comment (aryszka) [inline handlers/tokeninfo/jwt/handler.go]: this else branch is hit also when err == nil but !token.Valid

File: handlers/tokeninfo/jwt/handler.go [hunk 5]
```diff
@@ -16,21 +18,26 @@ type jwtHandler struct {
...
+		registerError(tie)
+		tokeninfo.Error(w, tie)
+		log.Println(err)
 		return
 	}
 
 	w.Header().Set("Content-Type", "application/json")
 	w.WriteHeader(http.StatusOK)
 	json.NewEncoder(w).Encode(ti)
+	measureRequest(start, fmt.Sprintf("planb.tokeinfo.jwt.%s.requests", ti.Realm))
```
Comment (aryszka) [inline handlers/tokeninfo/jwt/handler.go]: this crashes with nil pointer when the result from h.validateToken(r) is nil, nil

File: keys/openidprovider.go [hunk 6]
```diff
@@ -3,29 +3,25 @@ package keys
...
-const defaultRefreshInterval = 30 * time.Second
-
-const OPENID_PROVIDER_CONFIGURATION_URL = "OPENID_PROVIDER_CONFIGURATION_URL"
-
-func newCachingOpenIdProviderLoader() KeyLoader {
-	u := os.Getenv(OPENID_PROVIDER_CONFIGURATION_URL)
-	kl := &cachingOpenIdProviderLoader{url: u, keyCache: NewCache()}
-	schedule(defaultRefreshInterval, kl.refreshKeys)
+func NewCachingOpenIdProviderLoader(u *url.URL) KeyLoader {
+	kl := &cachingOpenIdProviderLoader{url: u.String(), keyCache: NewCache()}
```
Comment (aryszka) [inline keys/openidprovider.go]: since now this is public, it would be nice to check if u is not nil. just cosmetics, though.

**Pull request merged**